apis/python/src/tiledbsc/soma.py

```diff
@@ -547,6 +547,99 @@ def from_soma_slice(
 
         return soma
 
+    @classmethod
+    def map(
+        cls,
+        somas: List[SOMA],
+        soma_callback,
+        data: Optional[Dict[str, Any]] = None,
+        *,
+        max_thread_pool_workers: Optional[int] = None,
+        parallelize: bool = True,
+    ) -> Dict[str, Any]:
+        # type: ignore[no-untyped-def]
+        """
+        Invokes the callback function/lambda on every SOMA in the list, and returns a dict
+        from SOMA name to return value of that callback.  If `data` is not `None`, it should be a
+        dict from SOMA name to something that should be passed as a callback argument.
+        Example use: invoke this once with a per-SOMA attribute-filter query and get back a list of
+        obs IDs per SOMA. Invoke this again with a lambda that does something with each SOMA's obs
+        IDs.
+        ```
+        obs_ids_per_soma = SOMA.map(
+            somas,
+            lambda soma: list(soma.obs.query(
+                query_string='cell_type == "pericyte cell"', attrs=["cell_type"]
+            ).index)
+        )
+        X_dfs = SOMA.map(
+            somas,
+            lambda soma, obs_ids: soma.X.data.df(obs_ids=obs_ids), obs_ids_per_soma
+        )
+        ```
+        """
+        if parallelize:
+            return cls._map_in_parallel(
+                somas,
+                soma_callback,
+                data,
+                max_thread_pool_workers=max_thread_pool_workers,
+            )
+        else:
+            return cls._map_sequentially(somas, soma_callback, data)
+
+    # type: ignore[no-untyped-def]
+    @classmethod
+    def _map_in_parallel(
+        cls,
+        somas: List[SOMA],
+        soma_callback,
+        data: Optional[Dict[str, Any]] = None,
+        *,
+        max_thread_pool_workers: Optional[int] = None,
+    ) -> Dict[str, Any]:
+        """
+        Helper method for `map`.
+        """
+
+        if max_thread_pool_workers is None:
+            max_thread_pool_workers = SOMAOptions().max_thread_pool_workers
+
+        futures = {}
+        with ThreadPoolExecutor(max_workers=max_thread_pool_workers) as executor:
+            for soma in somas:
+                if data is None:
+                    futures[soma.name] = executor.submit(
+                        soma_callback,
+                        soma,
+                    )
+                else:
+                    futures[soma.name] = executor.submit(
+                        soma_callback,
+                        soma,
+                        data[soma.name],
+                    )
+
+        output = {}
+        for name, future in futures.items():
+            output[name] = future.result()
+        return output
+
+    @classmethod
+    def _map_sequentially(
+        cls,
+        somas: List[SOMA],  # type: ignore[no-untyped-def]
```

Comment:
i've spent two hours of my life, which i'll never get back, struggling and failing to find a syntax that `mypy` will accept for these _incredibly simple_ callbacks -- struggling and failing to find an acceptable syntax using `Callable` as well as struggling and failing to find an acceptable ignore-no-untyped-def syntax.

I'm open to constructive, non-dismissive, supportive feedback here from anyone who has it -- thank you in advance -- 🙏